configserver/src/main/java/com/yahoo/vespa/config/server/deploy/Deployment.java

```diff
@@ -165,6 +175,31 @@ private void restartServices(ApplicationId applicationId) {
         }
     }
 
+    private void waitForConfigToConverge(ApplicationId applicationId) {
+        BooleanFlag verify = Flags.CHECK_CONFIG_CONVERGENCE_BEFORE_RESTARTING.bindTo(applicationRepository.flagSource());
+        if ( ! verify.value()) return;
+
+        Instant end = clock.instant().plus(Duration.ofMinutes(10));
+        // Timeout per service when getting config generations
+        Duration timeout = Duration.ofSeconds(10);
+        do {
+            Application app = applicationRepository.getActiveApplication(applicationId);
+            log.info("Wait for services in " + applicationId + " to converge on new generation before restarting");
+            ConfigConvergenceChecker convergenceChecker = applicationRepository.configConvergenceChecker();
+            ServiceListResponse response = convergenceChecker.getConfigGenerationsForAllServices(app, timeout);
+            if (response.converged) {
+                log.info("services converged on new generation " + response.currentGeneration);
+                return;
+            } else {
+                log.info("services not converged on new generation, wanted generation: " + response.wantedGeneration +
+                                 ", current generation: " + response.currentGeneration + ", will retry");
+                try { Thread.sleep(10_000); } catch (InterruptedException e) { /* ignore */ }
+            }
+        } while (clock.instant().isBefore(end));
+
+        throw new RuntimeException("Config has not converged");
```

Comment:
I suppose this'll ultimately lead to a failed deployment, not retried by the controller during deployment jobs. Should we retry? If so, we need some way to communicate this.

Reply:
Yes, will look into it